controller/linodecluster_controller.go

```diff
@@ -18,45 +18,216 @@ package controller
+	clusterScope.LinodeCluster.Status.FailureReason = util.Pointer(failureReason)
+	clusterScope.LinodeCluster.Status.FailureMessage = util.Pointer(err.Error())
+
+	conditions.MarkFalse(clusterScope.LinodeCluster, clusterv1.ReadyCondition, string(failureReason), clusterv1.ConditionSeverityError, "%s", err.Error())
+
+	lcr.Recorder.Event(clusterScope.LinodeCluster, corev1.EventTypeWarning, string(failureReason), err.Error())
+}
+
+func (r *LinodeClusterReconciler) reconcileCreate(ctx context.Context, clusterScope *scope.ClusterScope, logger logr.Logger) error {
+	linodeNB, err := services.CreateNodeBalancer(ctx, clusterScope, logger)
+	if err != nil {
+		setFailureReason(clusterScope, cerrs.CreateClusterError, err, r)
+
+		return err
+	}
+
+	clusterScope.LinodeCluster.Spec.Network.NodeBalancerID = linodeNB.ID
+
+	linodeNBConfig, err := services.CreateNodeBalancerConfig(ctx, clusterScope, logger)
+	if err != nil {
+		setFailureReason(clusterScope, cerrs.CreateClusterError, err, r)
+
+		return err
+	}
 
-	// TODO(user): your logic here
+	clusterScope.LinodeCluster.Spec.ControlPlaneEndpoint = clusterv1.APIEndpoint{
+		Host: *linodeNB.IPv4,
+		Port: int32(linodeNBConfig.Port),
+	}
 
-	return ctrl.Result{}, nil
+	return nil
+}
+
+func (r *LinodeClusterReconciler) reconcileDelete(ctx context.Context, logger logr.Logger, clusterScope *scope.ClusterScope) error {
+	logger.Info("deleting cluster")
+	if clusterScope.LinodeCluster.Spec.Network.NodeBalancerID == 0 {
+		logger.Info("NodeBalancer ID is missing, nothing to do")
+		controllerutil.RemoveFinalizer(clusterScope.LinodeCluster, infrav1.GroupVersion.String())
+
+		return nil
+	}
+
+	if err := clusterScope.LinodeClient.DeleteNodeBalancer(ctx, clusterScope.LinodeCluster.Spec.Network.NodeBalancerID); err != nil {
+		logger.Info("Failed to delete Linode NodeBalancer", "error", err.Error())
+
+		// Not found is not an error
+		apiErr := linodego.Error{}
+		if errors.As(err, &apiErr) && apiErr.Code != http.StatusNotFound {
+			setFailureReason(clusterScope, cerrs.DeleteClusterError, err, r)
+
+			return err
+		}
+	}
+
+	conditions.MarkFalse(clusterScope.LinodeCluster, clusterv1.ReadyCondition, clusterv1.DeletedReason, clusterv1.ConditionSeverityInfo, "Load balancer deleted")
+
+	clusterScope.LinodeCluster.Spec.Network.NodeBalancerID = 0
+	controllerutil.RemoveFinalizer(clusterScope.LinodeCluster, infrav1.GroupVersion.String())
+
+	return nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *LinodeClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&infrastructurev1alpha1.LinodeCluster{}).
-		Complete(r)
+	controller, err := ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1.LinodeCluster{}).
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(mgr.GetLogger(), r.WatchFilterValue)).
+		Build(r)
+	if err != nil {
+		return fmt.Errorf("failed to build controller: %w", err)
+	}
+
+	return controller.Watch(
```

Comment:
I think similar to above we'll want to do something like below to add context around the error instead of just returning it as-is:
```suggestion
	if err = controller.Watch(
		...
	); err != nil {
		return fmt.Errorf("failed adding a watch for ready clusters": %w", err)
	}
	
	return nil
```